initialize JsonPackage lazily, don't eagerly resolve modules

jvasileff opened the pull request:
Previously:
1. When loading a module, imports were eagerly resolved and loaded
2. When loading a module, declarations were eagerly loaded for all
   packages.

This patch adds LazyPackage, allowing #2 to be delayed, which in turn
allows #1 to be skipped (it's ok, the typechecker will eventually
request all necessary modules to be loaded.)

The result is support for the following compile time scenarios:
1. Import of modules with circular dependencies (#6003)
2. Import of a module that depends on a module currently being compiled
   (a normally impossible scenario, but may be useful in some cases,
   such as compiling a language module that depends on an interop
   module that depends on the language module.)
3. The bizarre scenario of compiling a module that imports a module that
   is not currenlty being compiled that imports a third module that is
   currently being compiled (#5976)

This patch fixes #5976 and fixes the compile-time component of #6003.

jvasileff committed:
initialize JsonPackage lazily, don't eagerly resolve modules

Previously:

1. When loading a module, imports were eagerly resolved and loaded
2. When loading a module, declarations were eagerly loaded for all
   packages.

This patch adds LazyPackage, allowing #2 to be delayed, which in turn
allows #1 to be skipped (it's ok, the typechecker will eventually
request all necessary modules to be loaded.)

The result is support for the following compile time scenarios:

1. Import of modules with circular dependencies (#6003)
2. Import of a module that depends on a module currently being compiled
   (a normally impossible scenario, but may be useful in some cases,
   such as compiling a language module that depends on an interop
   module that depends on the language module.)
3. The bizarre scenario of compiling a module that imports a module that
   is not currenlty being compiled that imports a third module that is
   currently being compiled (#5976)

This patch fixes #5976 and fixes the compile-time component of #6003.

quintesse commented:
ok to test

jvasileff commented:
This probably needs more testing to make sure all necessary transitive dependencies are pulled in.

jvasileff commented:
WTF. Is this a 3+ year-old bug?

https://github.com/ceylon/ceylon/blob/1.2.1/typechecker/src/com/redhat/ceylon/compiler/typechecker/analyzer/ModuleValidator.java#L215

After the first circular dependency is detected (like, say, the language module import...), the rest of `moduleImports` are ignored.

Seems like it should be `continue;`, not `return;`.

quintesse commented:
> WTF? Is this a 3+ year-old bug?

Nobody uses `ceylon js`, it's just there to keep @chochos off the street ;)

chochos commented:
Well certainly nobody uses it like @jvasileff he always finds the weirdest bugs

jvasileff committed:
don't go running away at the first sight of a circular dependency

Processing of imports in verifyModuleDependencyTree() for a module was
aborted upon finding a previously visited import (circular dependency).

But, it's entirely possible for additional imports to exist that have no
yet been visited. This patch replaces 'return' with 'continue' to ensure
all imports are processed.

For example, with the following three modules:

    - A imports B
    - B imports A
    - B imports X

processing of module B's imports { A, X } would return at module A,
leaving module X unprocessed.

jvasileff committed:
avoid stackoverflow when searching for packages

Keep track of visited modules in getPackage() to avoid infinite
recursion

jvasileff commented:
The WTF code was in the typechecker (backend agnostic). The JVM backend was somehow immune, and the JS backend, before this PR, eagerly loaded modules on its own and was therefore also immune.

Anyway, I think this PR is ready. It wouldn't hurt for someone to review the changes but I think the biggest risk of breakage would be due to other bugs being exposed by these fixes. And, for the typechecker fix, the previous behavior was somewhat random, depending on the order imports were listed in `module.ceylon`.

FroMage commented:
Is this supposed to come in 1.2.2?

jvasileff commented:
I don't especially like last second merges so I'm fine waiting until right after the release. For my Dart use case, there is a workaround.

jvasileff commented:
@FroMage, @chochos, can this be merged now that 1.2.2 is out?

FroMage commented:
Fine by me, as long as the tiny change in `ModuleValidator` doesn't break any test.

FroMage commented:
Up to @chochos to merge it, though.

chochos commented:
Isn't there already a `LazyPackage` class in the JVM compiler? If this is the same one (just copied over), shouldn't it rather be in some library that's common to both compilers (ceylon-model perhaps)?

jvasileff commented:
The one for the JVM is JVM specific and not suitable for use here.
